### PR TITLE
[DataFrame] Added Implementations for equals, query, and some other operations

### PR DESCRIPTION
**Pull request by kunalgosar**
## What do these changes do?

The APIs for these implementations conform to that of pandas, many of the other operators rely on mapped calls to pandas functions. 

This PR implements:

1. equals
2. query
3. __ iter__
4. __ abs__
5. __ eq__
6. __ ne__
7. __ neg__

### CODE REVIEW

File: python/ray/dataframe/test/test_dataframe.py [hunk 1]
```diff
@@ -1224,10 +1248,33 @@ def test_quantile():
 
 
 def test_query():
-    ray_df = create_test_dataframe()
+    def generate_data():
```
Comment (devin-petersohn) [inline python/ray/dataframe/test/test_dataframe.py]: I prefer these be in `@pytest.fixture` calls and pass the different dataframes to it.

File: python/ray/dataframe/dataframe.py [hunk 2]
```diff
@@ -192,6 +201,15 @@ def _map_partitions(self, func, index=None):
 
         return DataFrame(new_df, self.columns, index=index)
 
+    def _update_inplace(self, new_dfs):
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: If we're adding this convenience method (which I agree is a good idea) we will also need to update `columns` and `index`.

File: python/ray/dataframe/dataframe.py [hunk 3]
```diff
@@ -72,6 +75,12 @@ def _default_index(self):
 
     index = property(_get_index, _set_index)
 
+    def _update_lengths(self):
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Prefer name change to `_compute_lengths()`. We have a set and the name doesn't sound much different.

File: python/ray/dataframe/dataframe.py [hunk 4]
```diff
@@ -192,6 +201,15 @@ def _map_partitions(self, func, index=None):
 
         return DataFrame(new_df, self.columns, index=index)
 
+    def _update_inplace(self, new_dfs):
+        """Updates the current DataFrame inplace
+        """
+        assert(len(new_dfs) > 0)
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: DataFrames can be empty. This is fine to leave for now, but we need to handle this case (also above in the constructor)

File: python/ray/dataframe/test/test_dataframe.py (hunk 1 above)
Comment (devin-petersohn) [reply python/ray/dataframe/test/test_dataframe.py]: You can also pass in what you want to query for.

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
This is great! Just a few comments and minor changes.

File: python/ray/dataframe/test/test_dataframe.py (hunk 1 above)
Comment (kunalgosar) [reply python/ray/dataframe/test/test_dataframe.py]: Done.

**Review by devin-petersohn — APPROVED** [review]
Looks great!

**Comment by devin-petersohn** [comment]
Passed on private-travis. OK to merge. Thanks @kunalgosar!

**Pull request merged**